src/recovery/mod.rs

```diff
@@ -1195,12 +1307,30 @@ where
             // epoch; we only modify state carefully in epoch order
             // once we know we won't be getting any input on closed
             // epochs.
-            let mut logic_cache: HashMap<StateKey, L> = key_to_resume_state_bytes
-                .into_iter()
-                .map(|(key, logic_resume_state_bytes)| {
-                    (key, logic_builder(Some(logic_resume_state_bytes)))
-                })
-                .collect();
+            let mut current_logic: HashMap<StateKey, L> = HashMap::new();
+            // Next awaken timestamp for each key. There is only a
+            // single awake time for each key, representing the next
+            // awake time.
+            let mut current_next_awake: HashMap<StateKey, DateTime<Utc>> = HashMap::new();
+
+            for (
+                key,
+                State {
+                    snapshot,
+                    next_awake,
+                },
+            ) in resume_state
+            {
+                current_logic.insert(key.clone(), logic_builder(Some(snapshot)));
+                match next_awake {
+                    Some(next_awake) => {
+                        current_next_awake.insert(key.clone(), next_awake);
+                    }
+                    None => {
+                        current_next_awake.remove(&key);
+                    }
+                }
+            }
```

Comment:
may I suggest this?
```rust
            let mut current_logic = HashMap::new();
            let mut current_next_awake = HashMap::new();

            for (key, state) in state_cache.into_iter() {
                current_logic.insert(key.clone(), logic_builder(Some(state.state_bytes)));

                if let Some(next_awake) = state.next_awake {
                    current_next_awake.insert(key.clone(), next_awake);
                } else {
                    current_next_awake.remove(&key);
                }
            }
```

Reply:
I want to explicitly destructure `State` so that if we add more state fields, we get a compiler error reminding us to unpack the new state into the operator. But I'll make it less of a giant line. And use `if let Some`.